src/server.c

```diff
@@ -939,6 +940,76 @@ void homekit_server_on_identify(client_context_t *context) {
     }
 }
 
+const char* homekit_generate_setupURI(homekit_server_t *server) {
+
+  const char* setupID = server->config->setupId;
+  char *setupCode = malloc(9);
+  setupCode[0] = server->config->password[0];
+  setupCode[1] = server->config->password[1];
+  setupCode[2] = server->config->password[2];
+
+  setupCode[3] = server->config->password[4];
+  setupCode[4] = server->config->password[5];
+
+  setupCode[5] = server->config->password[7];
+  setupCode[6] = server->config->password[8];
+  setupCode[7] = server->config->password[9];
+  setupCode[8] = 0;
+
+  int code = atoi(setupCode);
+  free(setupCode);
+
+  // Payload description 45 bits
+  // V = Version - 3 bits
+  // R = Reserved - 4 bits
+  // C = Category - 8 bits
+  // F = Flags - 4 bits
+  // P = Pin - 26 bits
+  // VVVRRRRCCCCCCCCFFFFPPPPPPPPPPPPPPPPPPPPPPPPPP
+  unsigned long long payload = 0;
+
+  uint8_t version, flags, reserved, category;
+  homekit_accessory_t *accessory = server->config->accessories[0];
+  category = (accessory->category & 0xFF);
+  version = 0;
+  reserved = 0;
+  flags = 2; // 2=IP, 4=BLE, 8=IP_WAC
+
+  char *setupURICodedPayload = malloc(9);
+  memset(setupURICodedPayload, '0', 8);
+  setupURICodedPayload[9] = 0;
+
+  payload |= (version & 0x7);
+
+  payload <<= 4;
+  payload |= (reserved & 0xf);
+
+  payload <<= 8;
+  payload |= (category & 0xff);
+
+  payload <<= 4;
+  payload |= (flags & 0xf);
+
+  payload <<= 27;
+  payload |= (code & 0x7ffffff);
+
+  int index = 8;
+  while (payload > 0) {
+      int remainder = payload % 36;
+      payload /= 36;
+      if (index >= 0) {
+          setupURICodedPayload[index--] = base36Table[remainder];
+      }
+  }
+
+  char *setupURI = malloc(7+9+6);
```

Comment:
I would be more safe if that would automatically adjust based on payloads. Like in this case:
```c
int setupURISize = snprintf(NULL, 0, "X-HM://%s%s", setupURICodedPyaload, setupID);

char *setupURI = malloc(setupURISize);
snprintf(setupURI, setupURISize, "X-HM://%s%s", setupURICodedPyaload, setupID);

return setupURI;
```